File: js/캐시(2018 kakao).js
function solution(cacheSize, cities) {
    if (cacheSize === 0) {
        return 5 * cities.length;
    }
    
    let answer = 0;
    let arr = [];
    
    for (city of cities) {
        city = city.toLowerCase();
        
        let idx = arr.indexOf(city);
        
        // index가 -1 : 배열 내에 원소가 없음
        // 따라서 -1보다 크면 배열 내에 원소가 있음을 의미
        if (idx > -1) {
            arr.splice(idx, 1);
            answer += 1;
        } else {
            if (arr.length >= cacheSize) {
                // arr길이가 cacheSize보다 크거나 같으면 마지막 원소 제거
                arr.shift();
            }
            answer += 5;
        }
        arr.push(city);
    }
    return answer;
}

File: python/캐시(2018 kakao).py
# LRU 알고리즘 - 캐시 교체 알고리즘
# LRU를 구현하기 위해서는 캐시가 가득 차서 더이상 공간이 없을 때, 가장 오랫동안 참조되지 않은 페이지를 찾아서 없애주는 작업이 필요하다.
# 즉, 참조하는 값이 캐시 안에 없다면 가장 오래 전에 참조한 값은 없애고 현재의 값을 캐시에 새로 넣어준다.
# 참조하는 값이 이미 캐시에 존재한다면 해당 값을 캐시의 가장 최근 위치로 넣어준다.
# -> deque 사용

from collections import deque

def solution(cacheSize, cities):
    buffer = deque(maxlen=cacheSize)
    total = 0

    if cacheSize == 0:
        return 5 * len(cities)

    for city in cities:
        city = city.lower()

        if city in buffer:
            total += 1
            buffer.remove(city)
            buffer.append(city)
        else:
            total += 5
            buffer.append(city)
    return total

